### ISSUE TITLE
Problems with loading extensions in VB.Net app + fix

### ISSUE THREAD

**Issue by ghost** [DotSpatial/DotSpatial#1239]
*IF BUG, INCLUDE THIS PART:*

### Steps to reproduce

1. Created a DotSpatial application in vb.Net in VS 2017, Win 10 Pro, Any CPU, as per examples, including AppManager, Header control, Toolstrip, Dockmanager and all appropriately linked.  The app works fine

2. The _Plugins_ and _Windows_ Extensions_ directories are copied from the respective distributions to the \bin\Debug directories. E.g. the for 2.0 distribution there are 14 sub-directories under the Plugins directory and 9 sub-directories under the Windows Extensions directory.

3.making sure the AppManager directories are set to the appropriate Plugins and _Windows_ Extensions_ directories

4. All works fine until a call is made to AppManager1.LoadExtensions()

DotSpatial version:  1.9, and 2.0 (via Nuget and downloaded/recompiled

### Expected behaviour

AppManager1.LoadExtensions should load extensions

### Actual behaviour

A range of error messages are being generated, many pertaining to files that cannot be found, some to methods not found.  Especially the error that 'DotSpatial.Controls, Version 0.11.9.6, Culture=neutral .... cannot be loaded' confuses me a lot, Version 0.11.9.6??

Can you also include a screen shot?
Just a sample of messages: 
![image](https://user-images.githubusercontent.com/45922440/50117576-25dcf400-0299-11e9-9adf-2d931d808f3d.png)

![image](https://user-images.githubusercontent.com/45922440/50117613-38572d80-0299-11e9-98db-11d69ebc2b2b.png)

![image](https://user-images.githubusercontent.com/45922440/50117633-45741c80-0299-11e9-80b2-b4debe2fdf0e.png)

![image](https://user-images.githubusercontent.com/45922440/50117651-4e64ee00-0299-11e9-8065-ce9077d7775c.png)

![image](https://user-images.githubusercontent.com/45922440/50117687-689ecc00-0299-11e9-919b-c5c59c53ba4f.png)

![image](https://user-images.githubusercontent.com/45922440/50117711-75bbbb00-0299-11e9-9698-057edff43f3e.png)

**Comment by pan054** [comment]
I got a bit further with this issue, I believe.  Stepping through the code, the first error message is generated at the SatisfyImportsOnStartupExtension.Activate() sub. It  installs packages from NuGet such as DockManager.0.12.5.14, Measure.1.0.1185,MenuBar.1.0.1200 etc..  After that, the App.RefreshExtensions() triggers message boxes such as:
![error mess](https://user-images.githubusercontent.com/45922440/50131595-9b16ec00-02ce-11e9-9bcf-d7cad9a3daea.jpg)

I believe the 'Controls, Version 0.11.9.6 ....' is one of the dependencies of one of the NuGet packages.  So it seems that some 'set' NuGet packages are being loaded while there are already ready-to-go local plugins and extensions.  And this is in the situation that the DotSpatial objects come from locally compiles modules.  But I believe it is the same for Nuget-dowlnloaded ones.
So the next question is: what can be done about this?

**Comment by sindizzy** [comment]
Make sure that your project is at least the version of the .NET Framework of DotSpatial.

**Comment by sindizzy** [comment]
I would also test your project compiling for x86.

**Comment by sindizzy** [comment]
I had many problems with the default project. I opened it in Visual Studio 2017 and it compiled just fine. I think by default it uses 2015 but that default c# compiler is not correct.

**Comment by pan054** [comment]
The eight or so DS apps I've created (in VB.Net and with various DS versions and manually configures user interfaces in VS2017) are all working.  It is just the extensions part that seem to be very dependent on all kind of other files.   
Would you have any idea why this need for (and absence of) DotSpatial.Controls, Version=0.11.9.6?

**Comment by sindizzy** [comment]
Are the libraries in both Debug and Release? Can you run DemoMap?

**Comment by pan054** [comment]
Solution
If you run your DotSpatial VB.Net app without the right shell command (see later), the DS framework downloads a couple of NuGet packages into a newly created directory _C:\ProgramData\ your_app_name.exe_ and will try to load those packages.  However, those packages cause many problems resulting in many error messages about files that cannot be loaded (see my previous messages).  So, unless you delete that directory, it will keep causing problems, even if you have added the right Shell commands.
This VB.Net code (VS2017, .Net 4.5.2, DS 2.0 RC, Win 10 Pro) worked for me:
```
Imports System.ComponentModel.Composition

Public Class Form1
    <Export("Shell", GetType(ContainerControl))>
    Private Shared shell As ContainerControl
    Public Sub New()
        ' This call is required by the designer.
        InitializeComponent()
        If DesignMode Then Return
        shell = Me
        AppManager1.LoadExtensions()
    End Sub
End Class
```
If you forget the ‘Shared’ keyword or the Import, things will not (I cannot get them to) work in VB.Net. (VS2017, .Net 4.5.2, DS 2.0 RC, Win 10 Pro).
Also, I could not get the ‘empty UI ‘ (where the DS framework builds your UI) approach to work.  So I added the UI elements manually (no SpatialToolstrip, no SpatialHeaderControl), see screendump below.
Thanks to sindizzy for his support.

![image](https://user-images.githubusercontent.com/45922440/50253306-86089d00-0435-11e9-8331-b3f07b491061.png)

**Comment by danames** [comment]
It strikes me that we may be at a point where it would make sense to ditch
NuGet and simplify this process so that all the plugins are downloaded,
compiled and manually included by the developer.
**********************************************************************
Daniel P. Ames, Ph.D., P.E.
Professor, Civil & Environmental Engineering,Brigham Young University,
Provo, Utah, USA
President, International Environmental Modelling & Software Society
Editor-in-Chief, Environmental Modelling & Software
http://ceen.et.byu.edu/content/dan-ames
http://www.researchgate.net/profile/Daniel_Ames
http://scholar.google.com/citations?user=S0GUCeUAAAAJ

**********************************************************************

**Comment by pan054** [comment]
Based on my experience over the last week:
- As per Dan's suggestion, get rid of the NuGet loading altogether, it caused me almost a week of work to find/repair and it would simplify debugging considerably.
- Get rid of the AppManager automatic UI generation completely (my preference), or at least make it an option in the AppManager properties. It is still playing havoc in my current project!
- A Version.txt file in each distribution, stating version, build, target .Net framework, and compilation date would be helpful.  Especially in the Master Branch.  Some diagram with DS interdependencies would also be very useful.
- It has been said before (sorry), but there is in effect no programmers documentation (other than the C# API) for the DS basics. There used to be quite helpful MapWindows doco for VB.Net with code snippets. Maybe it could be used as a basis for DS doco (I know what the answer will be to this suggestion :) ).
- What is the difference between Windows Extensions (almost all called 'plugins') and plugins?


Have a relaxing break.

**Comment by jany-tenaj** [comment]
1. I would be in favor of ditching the auto download of the old dlls. No idea what it's for. I had my problems with it, too, in the beginning.
But I don't see a reason to ditch nuget completely. If you use the appveyor nuget releases from the main page you get all dlls from the base projects for the current master and can update them together. They also show which dlls they depend on and download them, too, when you download them.

2. You can either use the automatic UI generation or build your UI yourself as long as you make sure everything the system needs is there. So no to ditching the automatic UI.

3. Not sure what distributions you're talking about.

4. We had some diagrams that were way outdated and got removed because of that.

5. Yes there is no documentation, but if you have the time, you're welcome to write some. Otherwise it's currently way down on the list of things to fix.

6. Plugins that are inside the Windows Extensions folder work only on Windows systems. Plugins in the plugins folder are supposed to work on other systems, too.

**Comment by jany-tenaj** [comment]
@sindizzy Yes you're right giving new developers a starting point would be nice. I started to work on updating the examples when I had a little time over on my own a account. But then something came up and I had to stop. Basically I can do fixes I need for my own work when I'm at work. But I can't go and work on the examples. 

What do you mean by add stuff to the tutorial? You're welcome to work on it and make a pull request.
You can also work on bringing over the snippets and documentation. But it would have to be updated to work with 2.0.

Can you explain the trouble you had with 2.?

**Comment by sindizzy** [comment]
The tutorial is outdated and based on a prior version. I remember when i first started I had to search for current methods to get started with DotSpatial. Most of the snippets worked out of the box except maybe those that now are in GeoAPI. The problem is that the CodePlex framework is difficult to decipher. I would need help in getting it started. I would also vote to put this in the Wiki of the GitHub page. I don't have access to that and I think that's only available to a user with certain permissions.

When I first started with DS it was quite difficult because I felt I had to hunt down pieces of missing or outdated info. CodePlex was a good start but when it got transferred to GitHub a lot of the info that was readily available became harder to find.

One thing that I found difficult was when I downloaded the source code I didn't have the required Visual Studio installed on my laptop. By default it is opened with the latest you have installed. Well that version may not have the C# compiler required by the repo. It took me several days to figure out why the code would not compile. Coding in C# is not my primary so didn't know that some code only compiles with a specific version of C#. Had this been spelled out somewhere like on the main page requirements it would have saved a C# newbie like me a lot of time chasing down the source of the problem. I now am aware of it but maybe this info would developers more apt to contribute.

Another road block for me was the process by which to contribute. I think I understand it now but for someone who doesn't know the GitHub methodology it is quite a daunting task. I gave up several times but kept coming back hoping to learn. There is a link on the main page on how to contribute but I think a video specific to DS would be very helpful to developers. I can try to put something together based on my experience.

All these notes are just my thoughts as I love working with DS and want to promote it. The DS developer group has done a great job with the GIS framework. I use it on a daily basis for my company work.

**Comment by sindizzy** [comment]
Btw where are the tutorials located? I thought they were Word documents or PDFs. I cant find them in the source repo.

**Comment by jany-tenaj** [comment]
@sindizzy you could check with @danames. He might have them as word files. I know some of his people updated them at some point, but they were not in the repo. That's why I was experimenting with the other system which would have been easier to update from everyone.

If you do work on the documentation, I'll try to help as much as I can.

**Comment by pan054** [comment]
The Mapwindows documentation I mentioned can be found here: 
https://www.mapwindow.org/documentation/mapwingis4.9/getting_started.html 

Thanks for explaining the plugins vs the windows extensions, it had me puzzled for some time.

Responding to jany-tenaj’s contribution:
1.  My suggestion pertained to the ‘hidden’ Nuget loads by the AppManager, not to using Nuget as a distribution mechanism. The Nuget distribution mechanism was the source of mistakes in my case, mainly due to my own errors. I now much prefer the ‘local compilation’ route.  Compiling the VS solutions have worked perfectly for all distributions, every time, so it has proven to be a very dependable way, with the added bonus of having the correct version source code at hand when debugging.

2. In a well-documented situation where it is clear which bits and pieces the system needs and how to satisfy those needs, that could work.  In the current situation, where I still have to trial and error  my way through it, why not make it an explicit option in the AppManager’s properties to use the UI generator (or not)?  I wouldn’t have a clue how to add a simple button to an auto-generated toolbar for instance, so how useful is it, other than generating demo projects?

For my info on how contributing to DS works: is there a procedure or a person that determines what is in and what is out (like for instance dropping the toolbar from V1.9)?  Or is it whatever the latest contributor choses to do? How do you become a contributor?
Cheers

**Comment by sindizzy** [comment]
I have posted about my experiences with contributing here https://github.com/DotSpatial/DotSpatial/issues/1171. I currently have 4 pull requests in the queue and I assume when the project admins have time they will review the fixes and add them to the main branch. 

If you like you can reach out to me and I will give you more details on how I am modifying the source code and then creating pull requests.

As for the tutorials I found them in the CodePlex archive, although its a dump of a wiki.

**Comment by jany-tenaj** [comment]
On the main page is a link to the [contributing.md](https://github.com/DotSpatial/DotSpatial/blob/master/.github/CONTRIBUTING.md) which explains how you can contribute.

**Comment by pan054** [comment]
FYI: the issue of 'forced UI creation' seems to have been around (and confusing some of us) for some time, see 2016 post   [#686](https://github.com/DotSpatial/DotSpatial/issues/686). 

Would this issue be enough as a request _to please add an option to the AppManager to switch automatic UI generation on or off_?  Or do I have to submit that somewhere else?

**Comment by sindizzy** [comment]
I also have posted about it although I cant find the post at the moment. I think it is great that the UI is all created for the average user. However some of us are medium to advanced programmers and trying to integrate DotSpatial into an already existing app is quite challenging.

**Comment by jany-tenaj** [comment]
@pan054 it would be nice if you'd create an issue.

**Comment by pan054** [comment]
No worries, happy to do that jany-tenaj.  Can you please tell me where/how to do that?

**Comment by jany-tenaj** [comment]
Just click on the new issue button on top and write what should chage. If you mention this issue heree they will be lknked so we'll know where the thought is coming from.

**Comment by pan054** [comment]
As per jany-tenay's suggestion, I have submitted a change request (post #1255)

**Comment by hydromodeler** [comment]
Has this issue raised in the original post been resolved in an update?  I am attempting to use the GDAL extension 2.0 with DotSpatial 2.0 in my C# Windows Forms program.  DotSpatial (other than the extension) was added to the program with NuGet packages.  I followed the 12-step process described [here](https://stackoverflow.com/questions/29122130/raster-data-loading-from-filepath-in-dotspatial/56511418#56511418) to manually load the GDAL extension.  When I add the DotSpatial.Data.Rasters.GdalExtension reference to my project, I get an error upon starting a run with the same top dialog shown in the original post of this thread ("Could not load type DotSpatial.Controls.IMapSelfLoadLayer").
